python/pyarrow/_flight.pyx

```diff
@@ -0,0 +1,454 @@
+        return result
+
+
+cdef class FlightDataStream:
+    cdef:
+        unique_ptr[CFlightDataStream] stream
+
+
+cdef class RecordBatchStream(FlightDataStream):
+    def __init__(self, reader):
+        # TODO: we don't really expose the readers in Python.
+        # self.stream.reset(None)
+        pass
+
+
+cdef CStatus _get_flight_info(void* self, CFlightDescriptor c_descriptor,
+                              unique_ptr[CFlightInfo]* info):
+    """Callback for implementing Flight servers in Python."""
+    cdef FlightDescriptor descriptor = \
+        FlightDescriptor.__new__(FlightDescriptor)
+    try:
+        descriptor.descriptor = c_descriptor
+        result = (<object> self).get_flight_info(descriptor)
+        print(result)
+        if not result:
+            info.reset(NULL)
+            # TODO:
+        else:
+            # TODO:
+            pass
+        return CStatus_OK()
+    except NotImplemented as err:
+        return CStatus_NotImplemented(repr(err).encode('utf-8'))
+    except Exception as err:
+        return CStatus_UnknownError(repr(err).encode('utf-8'))
+    # TODO: KeyError
+
+
+cdef CStatus _do_put(void* self, unique_ptr[CFlightMessageReader] reader):
+    """Callback for implementing Flight servers in Python."""
+    cdef:
+        FlightRecordBatchReader py_reader = FlightRecordBatchReader()
+        FlightDescriptor descriptor = \
+            FlightDescriptor.__new__(FlightDescriptor)
+
+    descriptor.descriptor = reader.get().descriptor()
+    py_reader.reader.reset(reader.release())
+    try:
+        result = (<object> self).do_put(descriptor, py_reader)
+        return CStatus_OK()
+    except Exception as err:
+        print(err)
+        return CStatus_UnknownError(repr(err).encode('utf-8'))
+
+
+cdef CStatus _do_get(void* self, CTicket ticket,
+                     unique_ptr[CFlightDataStream]* stream):
+    """Callback for implementing Flight servers in Python."""
+    py_ticket = Ticket()
+    py_ticket.ticket = ticket.ticket
+    try:
+        result = (<object> self).do_get(py_ticket)
+        if not isinstance(result, FlightDataStream):
+            raise TypeError("FlightServerBase.do_get must return "
+                            "a FlightDataStream")
+        stream[0] = move((<FlightDataStream> result).stream)
+        return CStatus_OK()
+    except Exception as err:
+        print(err)
+        return CStatus_UnknownError(repr(err).encode('utf-8'))
+
+
+cdef class FlightServerBase:
+    """A Flight service definition."""
+
+    cdef:
+        unique_ptr[PyFlightServer] server
+
+    def run(self, port):
+        # TODO: share one vtable between instances
+        cdef:
+            PyFlightServerVtable vtable = PyFlightServerVtable()
+        vtable.get_flight_info = &_get_flight_info
+        vtable.do_put = &_do_put
+        vtable.do_get = &_do_get
+        self.server.reset(new PyFlightServer(self, vtable))
+        self.server.get().Run(port)
```

Comment:
Should certainly run without the GIL, no?

Reply:
Ah good point. One thing I also want to circle back to is this breaks Ctrl-C on the Python side currently.